src/perf/event.cpp

```diff
@@ -0,0 +1,658 @@
+                 << ",config1=" << attr_.config1 << std::dec << std::noshowbase << "/";
+
+    set_scale(read_file_or_else<double>(event_path.replace_extension(".scale"), 1.0));
+    set_unit(read_file_or_else<std::string>(event_path.replace_extension(".unit"), "#"));
+
+    if (!event_is_openable())
+    {
+        throw EventProvider::InvalidEvent(
+            "Event can not be opened in process- or system-monitoring-mode");
+    }
+}
+
+void SysfsEvent::use_sampling_options(const bool& use_pebs, const bool& sampling,
+                                      const bool& enable_cct)
+{
+    if (use_pebs)
+    {
+        attr_.use_clockid = 0;
+    }
+
+    if (sampling)
+    {
+        Log::debug() << "using sampling event \'" << name_ << "\', period: " << attr_.sample_period;
+
+        attr_.mmap = 1;
+    }
+    else
+    {
+        // Set up a dummy event for recording calling context enter/leaves only
+        attr_.type = PERF_TYPE_SOFTWARE;
+        attr_.config = PERF_COUNT_SW_DUMMY;
+    }
+
+    attr_.sample_id_all = 1;
+    // Generate PERF_RECORD_COMM events to trace changes to the command
+    // name of a task.  This is used to write a meaningful name for any
+    // traced thread to the archive.
+    attr_.comm = 1;
+    attr_.context_switch = 1;
+
+    // TODO see if we can remove remove tid
+    attr_.sample_type |= PERF_SAMPLE_IP | PERF_SAMPLE_TID | PERF_SAMPLE_CPU;
+    if (enable_cct)
+    {
+        attr_.sample_type |= PERF_SAMPLE_CALLCHAIN;
+    }
+
+    attr_.precise_ip = 3;
+
+    // make event available if possible
+    update_availability();
+}
+
+EventGuard Event::open(std::variant<Cpu, Thread> location, int cgroup_fd)
+{
+    return EventGuard(*this, location, -1, cgroup_fd);
+}
+
+EventGuard Event::open(ExecutionScope location, int cgroup_fd)
+{
+    if (location.is_cpu())
+    {
+        return EventGuard(*this, location.as_cpu(), -1, cgroup_fd);
+    }
+    else
+    {
+        return EventGuard(*this, location.as_thread(), -1, cgroup_fd);
+    }
+}
+
+EventGuard Event::open_as_group_leader(ExecutionScope location, int cgroup_fd)
+{
+    attr_.read_format |= PERF_FORMAT_GROUP;
+    attr_.sample_type |= PERF_SAMPLE_READ;
+
+    return open(location, cgroup_fd);
+}
+
+EventGuard EventGuard::open_child(Event child, ExecutionScope location, int cgroup_fd)
+{
+    if (location.is_cpu())
+    {
+        return EventGuard(child, location.as_cpu(), fd_, cgroup_fd);
+    }
+    else
+    {
+        return EventGuard(child, location.as_thread(), fd_, cgroup_fd);
+    }
+}
+
+EventGuard::EventGuard() : fd_(-1)
+{
+}
+
+EventGuard::EventGuard(Event& ev, std::variant<Cpu, Thread> location, int group_fd, int cgroup_fd)
+: ev_(ev)
+{
+    // can be deleted when scope gets replaced
+    ExecutionScope scope;
+    std::visit(overloaded{ [&](Cpu cpu) { scope = cpu.as_scope(); },
+                           [&](Thread thread) { scope = thread.as_scope(); } },
+               location);
```

Comment:
```suggestion
    std::visit([&scope](auto loc) { scope = loc.as_scope(); },
               location);
```

`overloaded` is not needed in this case. Just use a simply lambda instead.